server/sonar-server/src/main/java/org/sonar/server/settings/ThreadLocalSettings.java

```diff
@@ -0,0 +1,147 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2016 SonarSource SA
+ * mailto:contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.server.settings;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Maps;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import org.sonar.api.CoreProperties;
+import org.sonar.api.ce.ComputeEngineSide;
+import org.sonar.api.config.Encryption;
+import org.sonar.api.config.PropertyDefinitions;
+import org.sonar.api.config.Settings;
+import org.sonar.api.server.ServerSide;
+
+import static com.google.common.base.Preconditions.checkState;
+
+/**
+ * Merge of {@link SystemSettings} and the global properties stored in the db table "properties". These
+ * settings do not contain the settings specific to a project.
+ *
+ * <p>
+ * System settings have precedence on others.
+ * </p>
+ *
+ * <p>
+ * The thread-local cache is optional. It is disabled when the method {@link #unload()} has not
+ * been called. That allows to remove complexity with handling of cleanup of thread-local cache
+ * on daemon threads (notifications) or startup "main" thread.
+ * </p>
+ */
+@ComputeEngineSide
+@ServerSide
+public class ThreadLocalSettings extends Settings {
+
+  private final Properties systemProps;
+  private final ThreadLocal<Map<String, String>> dbPropsThreadLocal = new ThreadLocal<>();
+  private SettingLoader settingLoader;
+
+  public ThreadLocalSettings(PropertyDefinitions definitions, Properties props) {
+    this(definitions, props, new NopSettingLoader());
+  }
+
+  @VisibleForTesting
+  ThreadLocalSettings(PropertyDefinitions definitions, Properties props, SettingLoader settingLoader) {
+    super(definitions, new Encryption(null));
+    this.settingLoader = settingLoader;
+    this.systemProps = props;
+
+    // TODO something wrong about lifecycle here. It could be improved
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1135)